### PR TITLE
[homekit] Allow configuring thermostat modes via metadata

### PR DESCRIPTION
**Pull request by ccutrer**
for example, if it's a water heater that's always in heat mode, so you don't have to create dummy items for the characteristics

this also adds the infrastructure necessary to use metadata characteristics as mandatary characteristics

it also adds TemperatureUnit as a characteristic the user can provide, and support TargetRelativeHumidity characteristic for Thermostat

### CODE REVIEW

**Review by yfre — APPROVED** [review]
good feature. 
pretty code heavy change for one feature, but you also did some code re-org to make cleaner. thank you. 

LGTM

File: bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java [hunk 1]
```diff
@@ -362,6 +365,23 @@ && isMandatoryCharacteristic(mainItem, characteristic)) {
                 }
             }
         }
+        mandatoryCharacteristics.forEach(c -> {
+            // Check every metadata key looking for a characteristics we can create
+            var config = mainItem.getConfiguration();
+            if (config == null)
+                return;
```
Comment (lsiepel) [inline bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java:372]:
```suggestion
            if (config == null) {
                return;
            }
```

File: bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java [hunk 2]
```diff
@@ -362,6 +365,23 @@ && isMandatoryCharacteristic(mainItem, characteristic)) {
...
+                    .collect(Collectors.toList())) {
+                var type = HomekitCharacteristicType.valueOfTag(entry.getKey());
+                if (type.isPresent() && isMandatoryCharacteristic(mainItem, type.get())) {
+                    var characteristic = HomekitMetadataCharacteristicFactory.createCharacteristic(type.get(),
+                            entry.getValue());
+
+                    characteristic.ifPresent(rc -> rawCharacteristics.add(rc));
+                }
+            }
+
```
Comment (lsiepel) [inline bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java:383]:
```suggestion
```

File: bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitMetadataCharacteristicFactory.java [hunk 3]
```diff
@@ -104,16 +106,23 @@ public class HomekitMetadataCharacteristicFactory {
...
+        if (type.isEmpty()) {
             return Optional.empty();
         }
-        return Optional.of(OPTIONAL.get(type.get()).apply(value));
+        return createCharacteristic(type.get(), value);
+    }
+
+    public static Optional<Characteristic> createCharacteristic(HomekitCharacteristicType type, Object value) {
+        if (!OPTIONAL.containsKey(type))
+            return Optional.empty();
```
Comment (lsiepel) [inline bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitMetadataCharacteristicFactory.java:124]:
```suggestion
        if (!OPTIONAL.containsKey(type)) {
            return Optional.empty();
        }
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks, some minor comments.

**Pull request merged**